Source/AwsCommonRuntimeKit/mqtt/Mqtt5.swift

```diff
@@ -0,0 +1,1053 @@
+    let negotiatedSettings: NegotiatedSettings
+
+    init (connackPacket: ConnackPacket, negotiatedSettings: NegotiatedSettings) {
+        self.connackPacket = connackPacket
+        self.negotiatedSettings = negotiatedSettings
+    }
+}
+
+/// Defines signature of the Lifecycle Event Connection Success callback
+typealias OnLifecycleEventConnectionSuccess = (LifecycleConnectSuccessData) -> Void
+
+/// Dataclass containing results of a Connect Failure Lifecycle Event.
+public class LifecycleConnectFailureData {
+
+    /// Error which caused connection failure.
+    let crtError: CRTError
+
+    /// Data model of an `MQTT5 CONNACK <https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901074>`_ packet.
+    let connackPacket: ConnackPacket
+
+    init (crtError: CRTError, connackPacket: ConnackPacket) {
+        self.crtError = crtError
+        self.connackPacket = connackPacket
+    }
+
+}
+
+/// Defines signature of the Lifecycle Event Connection Failure callback
+typealias OnLifecycleEventConnectionFailure = (LifecycleConnectFailureData) -> Void
+
+/// Dataclass containing results of a Disconnect Lifecycle Event
+public class LifecycleDisconnectData {
+
+    /// Error which caused disconnection.
+    let crtError: CRTError
+
+    /// Data model of an `MQTT5 DISCONNECT <https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901205>`_ packet.
+    let disconnectPacket: DisconnectPacket
+
+    init (crtError: CRTError, disconnectPacket: DisconnectPacket) {
+        self.crtError = crtError
+        self.disconnectPacket = disconnectPacket
+    }
+}
+
+/// Defines signature of the Lifecycle Event Disconnection callback
+typealias OnLifecycleEventDisconnection = (LifecycleDisconnectData) -> Void
+
+/// Configuration for the creation of MQTT5 clients
+public class ClientOptions {
+    /// Host name of the MQTT server to connect to.
+    var hostName: String
+
+    /// Network port of the MQTT server to connect to.
+    var port: UInt32
+
+    /// The Client bootstrap used
+    var bootstrap: ClientBootstrap
+
+    /// The socket properties of the underlying MQTT connections made by the client or None if defaults are used.
+    var socketOptions: SocketOptions
+
+    /// The TLS context for secure socket connections. If None, then a plaintext connection will be used.
+    var tlsCtx: TLSContext
+
+    /// The (tunneling) HTTP proxy usage when establishing MQTT connections
+    var httpProxyOptions: HTTPProxyOptions?
+
+    // TODO WebSocket implementation
+    /// This callback allows a custom transformation of the HTTP request that acts as the websocket handshake. Websockets will be used if this is set to a valid transformation callback.  To use websockets but not perform a transformation, just set this as a trivial completion callback.  If None, the connection will be made with direct MQTT.
+    // var websocketHandshakeTransform: Callable[[WebsocketHandshakeTransformArgs], None] = None
```

Comment:
Is there any concern about `websocketHandshakeTransform`?

Reply:
I haven't dug into it yet. I was thinking about digging into it once we have normal connections succeeding. What concerns do you mean?